homeassistant/components/crownstone/light.py

```diff
@@ -0,0 +1,223 @@
+"""Support for Crownstone devices."""
+from __future__ import annotations
+
+from functools import partial
+import logging
+from typing import TYPE_CHECKING, Any
+
+from crownstone_cloud.cloud_models.crownstones import Crownstone
+from crownstone_cloud.const import (
+    DIMMING_ABILITY,
+    SWITCHCRAFT_ABILITY,
+    TAP_TO_TOGGLE_ABILITY,
+)
+from crownstone_cloud.exceptions import CrownstoneAbilityError
+from crownstone_uart import CrownstoneUart
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    SUPPORT_BRIGHTNESS,
+    LightEntity,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    ABILITY_STATE,
+    CROWNSTONE_INCLUDE_TYPES,
+    CROWNSTONE_SUFFIX,
+    CROWNSTONE_USB,
+    DOMAIN,
+    SIG_CROWNSTONE_STATE_UPDATE,
+    SIG_UART_STATE_CHANGE,
+)
+from .devices import CrownstoneDevice
+from .helpers import map_from_to
+
+if TYPE_CHECKING:
+    from .entry_manager import CrownstoneEntryManager
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up crownstones from a config entry."""
+    manager: CrownstoneEntryManager = hass.data[DOMAIN][config_entry.entry_id]
+
+    entities = []
+    crownstone_usb_sphere_id = None
+
+    # Look for a Crownstone USB and what sphere it belongs to
+    for sphere in manager.cloud.cloud_data:
+        for crownstone in sphere.crownstones:
+            if crownstone.type == CROWNSTONE_USB:
+                crownstone_usb_sphere_id = sphere.cloud_id
+
+    # Add Crownstone entities that support switching/dimming
+    for sphere in manager.cloud.cloud_data:
+        for crownstone in sphere.crownstones:
+            if crownstone.type in CROWNSTONE_INCLUDE_TYPES:
+                # Crownstone can communicate with Crownstone USB
+                if sphere.cloud_id == crownstone_usb_sphere_id:
+                    entities.append(CrownstoneEntity(crownstone, manager.uart))
+                # Crownstone can't communicate with Crownstone USB
+                else:
+                    entities.append(CrownstoneEntity(crownstone))
+
+    async_add_entities(entities)
+
+
+def crownstone_state_to_hass(value: int) -> int:
+    """Crownstone 0..100 to hass 0..255."""
+    return map_from_to(value, 0, 100, 0, 255)
+
+
+def hass_to_crownstone_state(value: int) -> int:
+    """Hass 0..255 to Crownstone 0..100."""
+    return map_from_to(value, 0, 255, 0, 100)
+
+
+class CrownstoneEntity(CrownstoneDevice, LightEntity):
```

Comment:
You'd benefit from:

https://developers.home-assistant.io/docs/core/entity#entity-class-or-instance-attributes

Reduces the LOC a little bit in this class.